alg/teca_temporal_reduction.cxx

```diff
@@ -0,0 +1,648 @@
+#include "teca_temporal_reduction.h"
+
+#include "teca_cartesian_mesh.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_metadata.h"
+#include "teca_calendar_util.h"
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+#define TECA_DEBUG
+
+using std::string;
+using std::vector;
+using std::cerr;
+using std::endl;
+using std::cos;
+
+// --------------------------------------------------------------------------
+void summation_operator::initialize(double fill_value)
+{
+    this->fill_value = fill_value;
+}
+
+// --------------------------------------------------------------------------
+void summation_operator::update(
+    p_teca_variant_array &out_array,
+    p_teca_variant_array &out_valid,
+    const_p_teca_variant_array &in_array,
+    const_p_teca_variant_array &in_valid)
+{
+    for (unsigned int i = 0; i < out_array->size(); ++i)
+    {
+
+        double array_in, array_out;
+        in_array->get(i, array_in);
+        out_array->get(i, array_out);
```

Comment:
This is not the best API to use for heavy data processing.

1. This API fetches a single value. When the data is on the GPU this will pull a single value across the PCI bus in each iteration of the loop.
2. Data access API on `teca_variant_array` are virtual. There are 2 performance issues with calling virtual methods inside loops. a) virtual methods are resolved at runtime and as a result the compiler cannot optimize them. b) Dispatch (a big if/else that casts to the actual data type) is done in every loop iteration.
3. This API cannot be used on the GPU. It always puts data on the CPU.

Rather than use the virtual single value data access API defined in `teca_variant_array` we should:

1. Apply dispatch once outside of the loop. This will give us a typed variant array instance (ie `teca_variant_array_impl<T>`.  We use `TEMPLATE_DISPATCH` macro for this. 
2. Once you have the typed array, call the appropriate accessibility method. For eg if the load balancer assigns this thread to a CPU the device will be -1, and we'd call `teca_variant_array_impl<T>::get_cpu_accessible`. A pointer to the data then can be obtained, inside the loop you will use the pointer to access the data as needed.

Dispatch is done outside of the loop, once before the loop begins. Accessing the data inside the loop uses a pointer instead of a virtual call. The compiler can optimize this type of calculation.

You can use the `teca_l2_norm` as a template for this process.